examples/vision/siamese.py

```diff
@@ -0,0 +1,345 @@
+  #  to make it integral multiple of `num_col`
+  #  simply set it equal to num_row * num_col 
+  to_show = num_row*num_col
+
+  # Plot the images
+  fig, axes = plt.subplots(num_row, num_col, figsize=(5,5))
+  for images, labels in dataset.take(1):
+    for i in range(to_show):
+
+        # If the number of rows is 1, the axes array is one-dimensional
+        if num_row == 1:
+          ax = axes[i%num_col]
+        else:
+          ax = axes[i//num_col, i%num_col]
+          
+        # images[0][i][:,:,0] -> because it is
+        # (28,28,1) and imshow takes (28,82)
+        ax.imshow(tf.concat([images[0][i],images[1][i]],axis=1), cmap='gray')
+        if test:
+          ax.set_title('True: {} | Pred: {:.5f}'.format(
+          labels[i],
+          predictions[i][0]))
+        else:
+          ax.set_title('Label: {}'.format(labels[i]))
+  if test:
+    plt.tight_layout(rect = (0,0,1.9,1.9 ), w_pad=0.0)
+  else:
+    plt.tight_layout(rect = (0,0,1.5,1.5))
+  plt.show()
+
+
+
+
+
+visualize(train_ds, to_show=3)
+
+"""
+## Define the model
+
+There will be two input layers, each leading to its own network, which
+produces embeddings. Lambda layer will merge them using
+[Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) and the
+merged layer will be fed to final network.
+"""
+
+
+# Provided two tensors t1 and t2
+# Euclidean distance = sqrt(sum(square(t1-t2)))
+def euclidean_distance(vects):
+    """
+    Parameter
+    ---------
+    vect  : list
+            list containing two tensors of same length
+
+    Return
+    ------
+    Tensor containing euclidean distance
+    (as floating point value) between vectors
+    """
+
+    x, y = vects
+    sum_square = K.sum(K.square(x - y), axis=1, keepdims=True)
+    return K.sqrt(K.maximum(sum_square, K.epsilon()))
+
+
+
+input = Input((28,28,1))
+x = tf.keras.layers.BatchNormalization()(input)
+x = Conv2D(4, (5,5), activation = 'tanh')(x)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Conv2D(16, (5,5), activation = 'tanh')(x)
+x = AveragePooling2D(pool_size = (2,2))(x)
+x = Flatten()(x)
+
+x = tf.keras.layers.BatchNormalization()(x)
+x = Dense(10, activation = 'tanh')(x)
+embedding_network = Model(input, x)
+
+
+input_1 = Input((28,28,1))
+input_2 = Input((28,28,1))
+
+# As mentioned above, Siamese Network share weights between
+# tower networks (sister networks). To allow this, we will use
+# same embedding network for both tower networks.
+tower_1 = embedding_network(input_1)
+tower_2 = embedding_network(input_2)
+
+merge_layer = Lambda(euclidean_distance)([tower_1, tower_2])
+normal_layer = tf.keras.layers.BatchNormalization()(merge_layer)
+output_layer = Dense(1, activation="sigmoid")(normal_layer)
+siamese = Model(inputs=[input_1, input_2], outputs=output_layer)
+
+# Contrastive loss = mean( (1-true_value) * square(prediction) +
+#                         true_value * square( max(margin-prediction, 0) ))
+def contrastive_loss(margin=1, y_true, y_pred):
```

Comment:
This would not compile -- you can't have a non default argument followed by a default argument. Please make sure your code compile.

Reply:
Ouchh, why do i even do that :grimacing: :sweat_smile: